statsmodels/doe/latin.py

```diff
@@ -0,0 +1,230 @@
+
+    return sample
+
+
+def optimal_design(dim, n_sample, bounds=None, start_design=None, niter=1,
+                   force=False, optimization=True):
+    """Optimal design.
+
+    Optimize the design by doing random permutations to lower the centered
+    discrepancy. If `optimization` is False, `niter` design are generated and
+    the one with lowest centered discrepancy is return. This option is faster.
+
+    Centered discrepancy based design show better space filling robustness
+    toward 2D and 3D subprojections. Distance based design better space filling
+    but less robust to subprojections.
+
+    Parameters
+    ----------
+    dim : int
+        Dimension of the parameter space.
+    n_sample : int
+        Number of samples to generate in the parametr space.
+    bounds : tuple or array_like ([min, k_vars], [max, k_vars])
+        Desired range of transformed data. The transformation apply the bounds
+        on the sample and not the theoretical space, unit cube. Thus min and
+        max values of the sample will coincide with the bounds.
+    start_design : array_like (n_samples, k_vars)
+        Initial design of experiment to optimize.
+    niter : int
+        Number of iteration to perform.
+    force : bool
+        If `optimization`, force *basinhopping* optimization. Otherwise
+        grid search is used.
+    optimization : bool
+        Optimal design using global optimization or random generation of
+        `niter` samples.
+    Returns
+    -------
+    sample : array_like (n_samples, k_vars)
+        Optimal Latin hypercube Sampling.
+
+    References
+    ----------
+    [1] Damblin et al., "Numerical studies of space filling designs:
+    optimization of Latin Hypercube Samples and subprojection properties",
+    Journal of Simulation, 2013.
+
+    """
+    global best_doe, best_disc
+    best_doe = start_design
+    best_disc = np.inf
+
+    if (bounds is None) and (best_doe is not None):
+            bounds = np.array([best_doe.min(axis=0), best_doe.max(axis=0)])
+    if optimization:
+        if best_doe is None:
+            best_doe = orthogonal_latin_hypercube(dim, n_sample, bounds)
+
+        def _perturb_best_doe(x, bounds):
+            """Perturbe the DoE and keep track of the best DoE.
+
+            Parameters
+            ----------
+            x : list of int
+                It is a list of:
+                    idx : int
+                        Index value of the components to compute
+            bounds : tuple or array_like ([min, k_vars], [max, k_vars])
+                Desired range of transformed data. The transformation apply the
+                bounds on the sample and not the theoretical space, unit cube.
+                Thus min and max values of the sample will coincide with the
+                bounds.
+
+            Returns
+            -------
+            discrepancy : float
+                Centered discrepancy.
+
+            """
+            global best_doe, best_disc
```

Comment:
Using globals is not a good design if we can avoid it.
For example, I think those two objects could be added as argument to the function and added to the `args` in the optimization.

Reply:
Fact is that I do not know how not to use global here as at every step of the basinhopping we need to use the current best result.